odbc_fdw.c

```diff
@@ -875,6 +888,205 @@ odbcGetTableSize(odbcFdwOptions* options, unsigned int *size)
+  SQLLEN StrLen_or_Ind;
+} DataBinding;
+
+typedef struct {
+  Oid serverOid;
+  DataBinding* tableResult;
+  SQLHSTMT stmt;
+  SQLCHAR schema;
+  SQLCHAR name;
+  SQLUINTEGER rowLimit;
+  SQLUINTEGER currentRow;
+} TableDataCtx;
+
+
+Datum odbc_tables_list(PG_FUNCTION_ARGS)
+{
+	SQLHENV env;
+	SQLHDBC dbc;
+	SQLHSTMT stmt;
+	SQLRETURN ret;
+  SQLUSMALLINT i;
+  SQLUSMALLINT numColumns = 5;
+  SQLUSMALLINT bufferSize = 1024;
+  SQLUINTEGER rowLimit;
+  SQLUINTEGER currentRow;
+  SQLRETURN retCode;
+
+  FuncCallContext *funcctx;
+  TupleDesc tupdesc;
+  TableDataCtx *datafctx;
+  MemoryContext oldcontext;
+  DataBinding* tableResult;
+  AttInMetadata *attinmeta;
+
+  if (SRF_IS_FIRSTCALL()) {
+    funcctx = SRF_FIRSTCALL_INIT();
+    MemoryContext oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+    datafctx = (TableDataCtx *) palloc(sizeof(TableDataCtx));
+    tableResult = (DataBinding*) palloc( numColumns * sizeof(DataBinding) );
+    
+    char *serverName = text_to_cstring(PG_GETARG_TEXT_PP(0));
+    int serverOid = oid_from_server_name(serverName);
+
+    rowLimit = PG_GETARG_INT32(1); 
+    currentRow = 0;
+
+    odbcFdwOptions options;
+    odbcGetOptions(serverOid, NULL, &options);
+    odbc_connection(&options, &env, &dbc);
+    SQLAllocHandle(SQL_HANDLE_STMT, dbc, &stmt);
+
+    for ( i = 0 ; i < numColumns ; i++ ) {
+      tableResult[i].TargetType = SQL_C_CHAR;
+      tableResult[i].BufferLength = (bufferSize + 1);
+      tableResult[i].TargetValuePtr = palloc( sizeof(char)*tableResult[i].BufferLength );
+    }
+
+    for ( i = 0 ; i < numColumns ; i++ ) {
+      retCode = SQLBindCol(stmt, i + 1, tableResult[i].TargetType, tableResult[i].TargetValuePtr, tableResult[i].BufferLength, &(tableResult[i].StrLen_or_Ind));
+    }
+
+    if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+      ereport(ERROR,
+          (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+           errmsg("function returning record called in context "
+             "that cannot accept type record")));
+
+    attinmeta = TupleDescGetAttInMetadata(tupdesc);
+
+    datafctx->serverOid = serverOid;
+    datafctx->tableResult = tableResult;
+    datafctx->stmt = stmt;
+    datafctx->rowLimit = rowLimit;
+    datafctx->currentRow = currentRow;
+    funcctx->user_fctx = datafctx;
+    funcctx->attinmeta = attinmeta;
+
+    MemoryContextSwitchTo(oldcontext);
+  }
+
+  funcctx = SRF_PERCALL_SETUP();
+
+  datafctx = funcctx->user_fctx; 
+  stmt = datafctx->stmt;
+  tableResult = datafctx->tableResult;
+  rowLimit = datafctx->rowLimit;
+  currentRow = datafctx->currentRow;
+  attinmeta = funcctx->attinmeta;
+
+  retCode = SQLTables( stmt, NULL, SQL_NTS, NULL, SQL_NTS, NULL, SQL_NTS, (SQLCHAR*)"TABLE", SQL_NTS );
+  if (SQL_SUCCEEDED(retCode = SQLFetch(stmt)) && (rowLimit == 0 || currentRow < rowLimit)) {
+    char       **values;
+    HeapTuple    tuple;
+    Datum        result;
+
+    values = (char **) palloc(2 * sizeof(char *));
```

Comment:
AFAIK the values part is stored in the [attinmeta]() which is a struct to:
```
attinmeta is for use when returning tuples (i.e. composite data types)
and is not used when returning base data types. It is only needed if
you intend to use BuildTupleFromCStrings() to create the return tuple.
```

The attinmeta object is initialized inside the `FuncCallContext` [here](https://github.com/CartoDB/odbc_fdw/pull/38/files#diff-5a248444c4ecbae1926fa2df045aa136R1047) which has a `multi_call_memory_ctx`.

In summary, all is running in the `multi_call_memory_ctx` but maybe I missed something